src/envs/behavior_options.py

```diff
@@ -588,61 +662,12 @@ def grasp_obj_at_pos(
         plan.append(new_hand_pos + list(hand_orn))
         hand_pos = new_hand_pos
 
-    # Setup two booleans to be used as 'memory', as well as
-    # a 'reversed' plan to be used for our option that's
-    # defined below. Note that the reversed plan makes a
-    # copy of the list instead of just assigning by reference,
-    # and this is critical to the functioning of our option
-    reversed_plan = list(reversed(plan))
-    plan_executed_forwards = False
-    tried_closing_gripper = False
-
-    def graspObjectOption(_state: State,
-                          env: BehaviorEnv) -> Tuple[Array, bool]:
-        nonlocal plan_executed_forwards
-        nonlocal tried_closing_gripper
-        done_bit = False
-        if (not plan_executed_forwards and not tried_closing_gripper):
-            # Step thru the plan to execute Grasping
-            # phases 1 and 2
-            ret_action = get_delta_low_level_hand_action(
-                env,
-                plan[0][0:3],
-                plan[0][3:6],
-                plan[1][0:3],
-                plan[1][3:6],
-            )
-            plan.pop(0)
-            if len(plan) == 1:
-                plan_executed_forwards = True
-
-        elif (plan_executed_forwards and not tried_closing_gripper):
-            # Close the gripper to see if you've gotten the
-            # object
-            ret_action = np.zeros(17, dtype=float)
-            ret_action[16] = 1.0
-            tried_closing_gripper = True
-
-        else:
-            # Grasping Phase 3: getting the hand back to
-            # resting position near the robot.
-            ret_action = get_delta_low_level_hand_action(
-                env,
-                reversed_plan[0][0:3],
-                reversed_plan[0][3:6],
-                reversed_plan[1][0:3],
-                reversed_plan[1][3:6],
-            )
-            reversed_plan.pop(0)
-            if len(reversed_plan) == 1:
-                done_bit = True
-
-        return ret_action, done_bit
-
     p.restoreState(state)
     p.removeState(state)
 
-    return graspObjectOption
+    print(f"PRIMITIVE: grasp {obj.name} success! Plan found with " +
+          f"continuous params {grasp_offset}. Executing now.")
+    return create_grasp_policy(plan)
```

Comment:
i see, so the idea will be if we're in model mode, we shouldn't use this policy, but rather just extract the final state from the plan? and teleport to it?

is that final state from the plan the same as this variable `state`? or is `state` the state before we screw things up with all this planning?

Reply:
so the idea in model mode will be actually that we'll call a function `create_grasp_model` similar to `create_grasp_policy` that will take in the `plan` and then teleport the world to the right end configuration (not exactly, since we'll be doing some extra legwork to make it more realistic so that we can be sure that if the option model works, the actual option policy execution will also work)

`state` is a saved state from before we screw things up with planning that we can use to reset the world before returning our function (that's why there's the `p.restoreState(state)` line before the `return`). The plan is actually a series of waypoints that might be for the hand, or for the body depending on what planning function is called, and so we can't simply return the last element, we have to do some work to teleport the robot (and objects it might be holding, etc.)